pkg/plumbing/config/exec_config.go

```diff
@@ -0,0 +1,113 @@
+package config
+
+import (
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/warptools/warpforge/pkg/formulaexec"
+	"github.com/warptools/warpforge/pkg/plotexec"
+	"github.com/warptools/warpforge/pkg/workspace"
+	"github.com/warptools/warpforge/wfapi"
+)
+
+// ExecConfig is warpforge's internal configuration required to run Exec
+// Consumers should not accept this struct directly but should accept an interface for the parts that they need.
+type ExecConfig struct {
+	State
+	// WorkspaceSet is the workspace stack set to be used when executing.
+	// This controls behavior such as memoization and host warehouses.
+	workspace.WorkspaceSet
+	// This is the root filesystem as an fs.FS
+	Fsys fs.FS
+}
+
+func (c *ExecConfig) BinPath() string {
+	// determine the path of the running executable
+	// other binaries (runc, rio) will be located here as well
+	path, ok := c.State.Env[EnvWarpforgePath]
+	if ok {
+		return path
+	}
+	return filepath.Dir(c.State.ExecutablePath)
+}
+
+func (c *ExecConfig) KeepRunDir() bool {
+	return c.envIsSet(EnvWarpforgeKeepRundir)
+}
+
+func (c *ExecConfig) RunPathBase() string {
+	if value, ok := c.State.Env[EnvWarpforgeRunPath]; ok {
+		return value
+	}
+	return c.TempDir
+}
+
+func (c *ExecConfig) WarehousePathOverride() *string {
+	value, ok := c.State.Env[EnvWarpforgeWarehouse]
+	if !ok {
+		return nil
+	}
+	return &value
+}
+
+func (c *ExecConfig) WorkingDirectory() string {
+	return c.State.WorkingDirectory
+}
+
+// NewConfig returns a warpforge configuration
+//
+// Errors:
+//
+//   - warpforge-error-searching-filesystem -- unable to load workspace stack
+//   - warpforge-error-serialization -- failed to get copy of global state
+func NewConfig() (*ExecConfig, error) {
+	state, err := NewState()
+	if err != nil {
+		return &ExecConfig{}, err
+	}
+	fsys := os.DirFS("/")
+	wss, err := workspace.FindWorkspaceStack(fsys, "", state.WorkingDirectory[1:])
+	if err != nil {
+		return &ExecConfig{}, err
+	}
+	return &ExecConfig{
+		State:        state,
+		WorkspaceSet: wss,
+		Fsys:         fsys,
+	}, nil
+}
+
+func (c *ExecConfig) envIsSet(key string) bool {
+	_, result := c.State.Env[key]
+	return result
+}
+
+func (c *ExecConfig) envOverride(key string) (string, bool) {
+	a, b := c.State.Env[key]
+	return a, b
+}
+
+func PlotExecConfig(cfg *ExecConfig) *plotexec.ExecConfig {
```

Comment:
The config package importing one of the thing-doing packages is the opposite direction I'd expect the package relationships to go.  A config package is usually going to land at the deepest parts of the package import graph.  So having it import much else, in my experience, usually provokes a cycle.

Apparently it hasn't produced a cycle yet, if this compiles successfully -- but I'd bet several shiny nickles going this direction will produce a cycle sooner or later as the code grows.

Reply:
Importing global state into thing-doing packages is precisely what we are trying to avoid. If this causes a cycle then you have in fact imported global state and should redefine your package API.

Reply:
It might be useful to see how I think the import graph should look.
![imports](https://user-images.githubusercontent.com/1406278/212522010-24d8f32e-15d6-4108-bf73-e1d3cd529c00.jpg)

Reply:
If you have the iterated version of this diagram and wanna post that here too, I think it did get much improved over time :)

Reply:
![architecture](https://user-images.githubusercontent.com/1406278/214678135-43a41ee5-b1ce-4b82-a76b-8d4a3853ff5a.png)

Here's the updated version based.